### PR TITLE
FAT-5: Integrate existing Karate tests with TestRail

### PR DESCRIPTION
**Pull request by oleksandr-yatsenko1**
Initial integration with TestRails.
https://issues.folio.org/browse/FAT-5

image
image
image
image
image

### CODE REVIEW

File: acquisitions/src/main/resources/domain/cross-modules/cross-modules-junit.feature [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    * table modules
+      | name                |
+      | 'mod-invoice'       |
+      | 'mod-finance'       |
+      | 'mod-orders'        |
+      | 'mod-login'         |
+      | 'mod-permissions'   |
+      | 'mod-configuration' |
+
+    * def testAdmin = {tenant: '#(testTenant)', name: 'test-admin', password: 'admin'}
```
Comment (KreiDer) [inline acquisitions/src/main/resources/domain/cross-modules/cross-modules-junit.feature:14]: remove testAdmin and testUser

File: acquisitions/src/main/resources/domain/mod-finance/finance-junit.feature [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    * url baseUrl
+    * table modules
+      | name                  |
+      | 'mod-finance'         |
+      | 'mod-login'           |
+      | 'mod-permissions'     |
+      | 'mod-finance-storage' |
+      | 'mod-configuration'   |
+
+    * def testAdmin = {tenant: '#(testTenant)', name: 'test-admin', password: 'admin'}
```
Comment (KreiDer) [inline acquisitions/src/main/resources/domain/mod-finance/finance-junit.feature:13]: remove this
 * def testAdmin = {tenant: '#(testTenant)', name: 'test-admin', password: 'admin'}
    * def testUser = {tenant: '#(testTenant)', name: 'test-user', password: 'test'}

File: acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature [hunk 3]
```diff
@@ -0,0 +1,28 @@
+Feature: mod-invoice integration tests
+
+  Background:
+    * def testAdmin = {tenant: '#(testTenant)', name: 'test-admin', password: 'admin'}
```
Comment (KreiDer) [inline acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature:4]: remove this
 def testAdmin = {tenant: '#(testTenant)', name: 'test-admin', password: 'admin'}
    * def testUser = {tenant: '#(testTenant)', name: 'test-user', password: 'test'}

File: acquisitions/src/main/resources/domain/mod-orders/features/encumbrance-tags-inheritance.feature [hunk 4]
```diff
@@ -24,6 +24,10 @@ Feature: Verify once order is opened or poline is updated, encumbrance inherit p
 
     * configure retry = { count: 4, interval: 1000 }
 
+    # set common data used by scenarios
+    * def sharedCallable = callonce read('classpath:common/shared-callable.feature')
```
Comment (KreiDer) [inline acquisitions/src/main/resources/domain/mod-orders/features/encumbrance-tags-inheritance.feature:28]: update this with GET orderline below

File: acquisitions/src/main/resources/domain/mod-orders/orders-junit.feature [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  Background:
+    * url baseUrl
+    * table modules
+      | name                |
+      | 'mod-orders'        |
+      | 'mod-login'         |
+      | 'mod-permissions'   |
+      | 'mod-configuration' |
+    * print '### testTenant', testTenant
```
Comment (KreiDer) [inline acquisitions/src/main/resources/domain/mod-orders/orders-junit.feature:11]: remove this
* print '### testTenant', testTenant
    * print '### testAdmin', testAdmin
    * print '### testUser', testUser

File: acquisitions/src/test/java/org/folio/CrossModulesApiTest.java [hunk 6]
```diff
@@ -1,14 +1,49 @@
...
-    runHook();
-    return Karate.run("classpath:domain/cross-modules/cross-modules.feature");
+  private static String TEST_BASE_PATH = "classpath:domain/cross-modules/features/";
+
+  public CrossModulesApiTest() {
+    super(TEST_BASE_PATH, TEST_SUITE_NAME, DEFAULT_SUITE_ID, DEFAULT_SECTION_ID);
   }
-}
+
+//  @Ignore
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/CrossModulesApiTest.java:22]: remove this comment code

File: acquisitions/src/test/java/org/folio/InvoicesApiTest.java [hunk 7]
```diff
@@ -1,14 +1,64 @@
...
+  public InvoicesApiTest() {
+    super(TEST_BASE_PATH, TEST_SUITE_NAME, DEFAULT_SUITE_ID, DEFAULT_SECTION_ID);
   }
-}
+
+//  @Ignore
+//  @Karate.Test
+//  Karate invoiceTest() {
+//    runHook();
+//    return Karate.run("classpath:domain/mod-invoice/invoice.feature");
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/InvoicesApiTest.java:26]: remove this comment code

File: acquisitions/src/test/java/org/folio/OrdersApiTest.java [hunk 8]
```diff
@@ -1,14 +1,64 @@
...
-    runHook();
-    return Karate.run("classpath:domain/mod-orders/orders.feature");
+  private static String TEST_BASE_PATH = "classpath:domain/mod-orders/features/";
+
+  public OrdersApiTest() {
+    super(TEST_BASE_PATH, TEST_SUITE_NAME, DEFAULT_SUITE_ID, DEFAULT_SECTION_ID);
+  }
+
+//  @Ignore
+//  @Karate.Test
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/OrdersApiTest.java:22]: remove this comment code

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+    logger.info("### runFeature result {} ", results.getErrorMessages() );
+  }
+
+  protected void runFeatureTest(String testFeatureName) throws IOException {
+    if (!testFeatureName.endsWith("feature")) {
+      testFeatureName = testFeatureName.concat(".feature");
+    }
+    Results results = Runner.path(basePath.concat(testFeatureName))
+        .tags("~@Ignore", "~@NoTestRail")
+        .parallel(1);
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:193]: add assertEquals to check failed cases

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+  protected static boolean isTestRailIntegrationEnabled() {
+    boolean isTestRailsEnabled = System.getProperty("testrail_url") != null;
+    logger.debug("TestRails integration status, isTestRailsEnabled: {}", isTestRailsEnabled);
+    return isTestRailsEnabled;
+  }
+
+  protected static void runFeature(String featurePath) {
+    Results results = Runner.path(featurePath)
+        .parallel(1);
+    logger.info("### runFeature result {} ", results.getErrorMessages() );
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:184]: add assertEquals to check failed cases

**Review by AndreiMPV — APPROVED** [review]

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+        deleteScenarios(existingScenarios);
+      }
+
+      for (Map.Entry<String, Results> results : resultsMap.entrySet()) {
+        System.out.printf("Updating cases for %s\n", results.getKey());
+        updateScenariosInTestSuite(results.getValue(), existingScenarios);
+        System.out.printf("Posting results for %s:\n", results.getKey());
+        postTestRunResults(results.getValue());
+      }
+//            JSONObject c = (JSONObject) client.sendPost("close_run/" + runId, new JSONObject());
```
Comment (KreiDer) [inline acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:242]: remove comment code

File: common/src/main/resources/common/shared-callable.feature [hunk 12]
```diff
@@ -0,0 +1,5 @@
+@ignore
+Feature: Commonly callable feature.
```
Comment (KreiDer) [inline common/src/main/resources/common/shared-callable.feature:2]: remove this feature file

**Review by KreiDer — CHANGES_REQUESTED** [review]

File: acquisitions/src/main/resources/domain/cross-modules/cross-modules-junit.feature (hunk 1 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/cross-modules/cross-modules-junit.feature:14]: Done

File: acquisitions/src/main/resources/domain/mod-finance/finance-junit.feature (hunk 2 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/mod-finance/finance-junit.feature:13]: Done

File: acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature (hunk 3 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature:4]: Done

File: acquisitions/src/main/resources/domain/mod-orders/features/encumbrance-tags-inheritance.feature (hunk 4 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/mod-orders/features/encumbrance-tags-inheritance.feature:28]: Done

File: acquisitions/src/main/resources/domain/mod-orders/orders-junit.feature (hunk 5 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/mod-orders/orders-junit.feature:11]: Done

File: acquisitions/src/test/java/org/folio/OrdersApiTest.java (hunk 8 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/OrdersApiTest.java:22]: Done

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java (hunk 9 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:193]: Done

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java (hunk 10 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:184]: Done

File: acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java (hunk 11 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/testrail/AbstractTestRailIntegrationTest.java:242]: Done

File: common/src/main/resources/common/shared-callable.feature (hunk 12 above)
Comment (oleksandr-yatsenko1) [reply common/src/main/resources/common/shared-callable.feature:2]: Done

File: acquisitions/src/test/java/org/folio/CrossModulesApiTest.java (hunk 6 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/CrossModulesApiTest.java:22]: Done

File: acquisitions/src/test/java/org/folio/InvoicesApiTest.java (hunk 7 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/test/java/org/folio/InvoicesApiTest.java:26]: Done

**Review by KreiDer — APPROVED** [review]

File: acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature (hunk 3 above)
Comment (oleksandr-yatsenko1) [reply acquisitions/src/main/resources/domain/mod-invoice/invoice-junit.feature]: Will remove it, thank you.

**Pull request merged**